Implementation of Elections spec

gordonje opened the pull request:
As defined in this [draft proposal](https://github.com/opencivicdata/docs.opencivicdata.org/blob/master/proposals/drafts/elections.rst).

gordonje commented:
@jamesturk given the movement toward separate apps, should spike this PR and submit another after PR #85 is merged?

Or I could also make a PR on the split-apps branch, and this could all go up together.

jamesturk commented:
let's keep these in separate PRs, but yeah I think we should wait to merge until we've decided one way or another on #83- I'll ping the relevant parties to see if they're OK with the plan

gordonje commented:
@jamesturk This one is ready for review again. Below are some notes on differences since I first put in the PR.

### Removed inheritance from non-abstract models
- `Election` no longer subclasses `Event`. There's been some discussion, though, about adding an abstract `EventBase` in `opencivicdata.core` someday.
	- In the mean time, I'm not clear on the proper format for `Election.id`. I currently have it as `ocd-election/...`, but should it be `ocd-event/...`?
- `ContestBase` is now an abstract model from which all the contest models inherit.
- `RetentionContest` no longer inherits from `BallotMeasureContest` but remains under the `contests.ballot_measure` namespace.

### Added models
- `ElectionSource`, which mirrors the `EventSource` model. I don't think any of the other event-related models have any relevance to elections.
- `RetentionContestOption`, which mirrors `BallotMeasureContestOption`.
- Split up `ContestIdentifier` into four contest-class specific models.
- Same as above for `ContestSource`.

### Removed models
- Replaced `Party` model with references to `Organization` from core (per issue [#89](https://github.com/opencivicdata/docs.opencivicdata.org/issues/89) from the specs repo). These foreign key relationships (e.g., `Candidacy.party`) are limited to only organizations with the classification "party", though I believe that's only forced on Django forms, not by the ORM or db.

jamesturk commented on .coveragerc:
```diff
@@ -1,6 +1,4 @@
-        opencivicdata/legislative/migrations/*
-        opencivicdata/legislative/admin/*
+        opencivicdata/*/migrations/*
+        opencivicdata/*/admin/*
```
👍

jamesturk commented on opencivicdata/elections/admin/candidacy.py:
```diff
@@ -0,0 +1,70 @@
+        'updated_at',
+    )
+    # date_hierarchy across relations was added to django 1.11
+    if django_version[0] >= 1 and django_version[1] >= 11:
```
I'm curious if you have a need for Django < 1.11 support?  technically this supports 1.10 right now, but since 1.11 is LTS I'm OK w/ new stuff requiring 1.11

gordonje replied:
Our package that will depend on `python-opencivicdata` currently supports Django>=1.9, but this particular feature isn't super important. Just adds a nicer filter option to this admin for users who are at least at Django 1.11.

jamesturk commented on opencivicdata/legislative/models/__init__.py:
```diff
@@ -6,4 +6,4 @@
 from .event import (Event, EventLocation, EventMedia, EventMediaLink, EventDocument, EventLink,
                     EventSource, EventParticipant, EventAgendaItem, EventRelatedEntity,
-                    EventAgendaMedia, EventAgendaMediaLink, EventDocumentLink)
+                    EventAgendaMedia, EventAgendaMediaLink, EventDocumentLink)
```
this might be a mistake?

gordonje replied:
Yeah, I'm not sure what happened exactly, as I would not have had any reason to change this file. Looks like the newline at the end of the file was deleted. I just added it back.

jamesturk commented on opencivicdata/elections/models/candidacy.py:
```diff
@@ -0,0 +1,117 @@
+        max_length=300,
+        help_text="For preserving the candidate's name as it was of the candidacy."
+    )
+    filed_date = models.DateField(
```
should this be a fuzzy date or will we always require Y-m-d

gordonje replied:
In our case, all of these dates are exact. I think the source for this data in most states will be a form the candidate files with an election authority. I think you'll either have the full date or no date at all.

But maybe we want to allow users to put in their best guess for the date? Or if the default for OCD is to use fuzzy dates unless there's a good reason to be exact, we would be fine with that.

jamesturk commented on opencivicdata/elections/models/election.py:
```diff
@@ -0,0 +1,90 @@
+        max_length=300,
+        help_text='Name of the Election.',
+    )
+    date = models.DateField(
```
this being an exact date makes sense to me

I think we agreed to sideline questions about multi-day elections (e.g. vote-by-mail)

gordonje replied:
IMO even vote by mail elections will want to be tied to a specific date on which the votes are expected to be counted and the results will be published. But, yeah, a convo for another time when we know users will need to track when ballots can be cast.

gordonje commented:
@jamesturk so I think the proposed changes are:
1. Drop the election date hierarchy filter on `CandidacyAdmin` so as not to introduce a Django >= 1.11 feature just yet.
2. Make `Candidacy.filed_date` a fuzzy date field, rather inheriting from `DateField` in `django.db.models`.

Should I go ahead and make either or both of these?

jamesturk commented:
- no need to drop the date hierarchy filter, definitely want to support Django >= 1.11 
- I'll defer to you on the filed_date issue, I don't know the data well enough to know if a fuzzy date is necessary, but if you think it'd be useful I'd switch it to that

I don't think I have any further changes, I think we could go ahead and accept this & consider it provisional (so a bit less strict on backwards incompatible changes until it stabilizes)

gordonje commented:
Cool! Let's do this.

palewire commented:
This is excellent news. Thrilled to see this merged and released.
